Tutorial: Getting started (#590)
tinacms/tinacms.org: content/docs/getting-started/edit-content.md

---
title: Editing Content
prev: /docs/getting-started/cms-set-up
next: /docs/getting-started/backends
---

> The purpose of a CMS is to allow editors to change content. [Forms](/docs/plugins/forms) are a fundamental part of this — providing the editing interface to edit data. In this step, we will **create and register a form to edit data** rendered on this page.

Reviewer: ```suggestion
The purpose of a CMS is to allow editors to change content. [Forms](/docs/plugins/forms) are a fundamental part of this as they define the editing interface for content. In this step, we will **create and register a form to edit data** rendered on this page.
```

## Create & Register a Form

We will use the `useForm` hook to [create the form](/docs/plugins/forms#creating-forms).

> Simply creating the form doesn't make it render in the sidebar, we also need to register it with the CMS. With the `usePlugin` hook, we'll [register the form](/docs/plugins/forms#registering-forms) so it renders in the sidebar.

Reviewer: ```suggestion
Simply creating the form doesn't make it render in the sidebar, we also need to [register it](/docs/plugins/forms#registering-forms) with the CMS. With the `usePlugin` hook, we'll registers the form so it is rendered in the sidebar.
```

    label: 'Edit Page',
    fields: [
      {
        name: 'title',
        label: 'Title',
        component: 'text',
      },
      {
        name: 'body',
        label: 'Body',
        component: 'textarea',
      },
    ],
    initialValues: pageData,
    onSubmit: async () => {
      window.alert('Saved!')
    },
  }

  // 3. Create the form
  const [editableData, form] = useForm(formConfig)

  // 4. Register it with the CMS
  usePlugin(form)

  return (
    <section className="App-header">
      <img src={logo} className="App-logo" alt="logo" />
      {/**
       * 5. Render the `editableData` returned from `useForm`
       */}
      <h1>{editableData.title}</h1>
      <p>{editableData.body}</p>
      <ExitButton />
    </section>
  )
}

//...
```

## Form Configuration

`useForm` needs a [form configuration object](/docs/plugins/forms#form-configuration) with properties that determine how the form behaves on load and save, what fields are available, and other metadata.

While there are other form config properties, the key ones to examine in the example above are `id`, `fields`, `initialValues`, and `onSubmit`. We'll look closely at each of these now.

### _id_

> The `id` is unique identifier for the form. Since our example is the only instance of this form, the `id` is hard-coded. But if this form were in a template component, we would generate the `id` from the content to distinguish it from other form instances.

Reviewer: ```suggestion
The `id` is unique identifier for the form. For this simple example the `id` is hard-coded, however it is good practice to set `id` to as a unique identifier from the content being edited.
```

### _fields_

The `fields` array is comprised of [field definitions](/docs/plugins/fields#field-definition). All fields share a common [base configuration](docs/plugins/fields#field-config). Field definitions need at least two properties: a `name` or the path to the data and a `component` to edit that data from.

You will be working with fields a lot in Tina. To get more familiar, try to adjust the `label` property or add a new field to the array. Reference [the documentation](/docs/plugins/fields) for a full list of default field plugins.

### _initialValues_

`initialValues` contains the data used to populate the form. If you do need to load data asynchronously on form creation, you'd use a function called `loadInitialValues` (we'll get to that soon).

### _onSubmit_

This is a function that runs when the form is saved. Right now it just gives an alert message, but we will use this function later to post changes to a backend.

## Edit the data

Head back to http://localhost:3000, enable the CMS, open the sidebar and try to update the content. You should now be able to edit the title and body copy on the demo!

> Next, we'll look at setting up a simple backend to retrieve and save data from a 3rd-party API.

Reviewer: I wanted to do this right away. What if we move this above the "Form Configuration" section

* Make the change
* Look at the change
* Explain the change

---
title: Editing Content
prev: /docs/getting-started/cms-set-up
next: /docs/getting-started/backends
---

The purpose of a CMS is to allow editors to change content. [Forms](/docs/plugins/forms) are a fundamental part of this as they define the editing interface for your content. In this step, we will **create and register a form to edit data** rendered on this page.

## Create & Register a Form

We will use the `useForm` hook to [create the form](/docs/plugins/forms#creating-forms).

> Simply creating the form doesn't make it render in the sidebar, we also need to [register it](/docs/plugins/forms#registering-forms) with the CMS. With the `usePlugin` hook, we'll register the form so it will be rendered in the sidebar.

Reviewer 2: ```suggestion
Calling `useForm` won't automatically make our form appear in the sidebar. There are multiple ways to use forms with Tina, and the sidebar is merely one option.

Although it doesn't automatically appear in the sidebar, the form returned by `useForm` is pre-configured to work as a sidebar plugin. To add this form to the sidebar, all we have to do is pass it into the `usePlugin` hook.
```

> <!-- Todo provide explanation of why these are two steps -->

Reviewer 2: ```suggestion
```

can remove if previous suggestion is accepted

> > _Fun-Fact:_ Forms are a type of [plugin](/docs/plugins) in Tina.

Reviewer 2: ```suggestion
> _Fun Fact:_ Forms are a type of [plugin](/docs/plugins) in Tina.
```

4. Register the form with `usePlugin`
5. Render the data returned from `useForm`

**src/App.js**

```js
// 1. Import `useForm` and `usePlugin`
import { TinaProvider, TinaCMS, useCMS, useForm, usePlugin } from 'tinacms'

//...

const pageData = {
  title: 'Tina is not a CMS',
  body: 'It is a toolkit for creating a custom CMS.',
}

function PageContent() {
  // 2. Define the form configuration object
  const formConfig = {
    id: 'tina-tutorial-index',
    label: 'Edit Page',
    fields: [
      {
        name: 'title',
        label: 'Title',
        component: 'text',
      },
      {
        name: 'body',
        label: 'Body',
        component: 'textarea',
      },
    ],
    initialValues: pageData,
    onSubmit: async () => {
      window.alert('Saved!')
    },
  }

  // 3. Create the form
  const [editableData, form] = useForm(formConfig)

  // 4. Register it with the CMS
  usePlugin(form)

  return (
    <section className="App-header">
      <img src={logo} className="App-logo" alt="logo" />
      {/**
       * 5. Render the `editableData` returned from `useForm`
       */}
      <h1>{editableData.title}</h1>
      <p>{editableData.body}</p>

>       <ExitButton />

Reviewer 2: ```suggestion
      <EditButton />
```

    </section>
  )
}

//...
```

## Edit the data

Head back to http://localhost:3000, enable the CMS, open the sidebar and try to update the content. You should now be able to edit the title and body copy on the demo!

![tina-tutorial-edit-data](/img/getting-started/edit-data.png)

## Form Configuration

To gain a little clarity into how this is working, let's look closer at the `formConfig`.

`useForm` needs a [form configuration object](/docs/plugins/forms#form-configuration) with properties that determine how the form behaves on load and save, what fields are available, and other metadata.

While there are other form config properties, the key ones to examine in the example above are `id`, `fields`, `initialValues`, and `onSubmit`. We'll look closely at each of these now.

### _id_

> The `id` is unique identifier for the form. For this simple example the `id` is hard-coded, however it is good practice to set `id` to as a unique identifier from the content being edited.

Reviewer 2: ```suggestion
The `id` is a **unique identifier** for the form. The `id` is hard-coded in our example, which is OK because the `PageContent` component is only used to render the home page. When you create a form inside of a component that is used on multiple pages, you will want to take care to ensure the form's `id` uniquely identifies the content currently being edited. Typically, you would use the content's filename, or a unique key returned from a content API.
```

I want to be a little stronger than just saying "good practice", as certain things will not work right if IDs aren't unique

      {/**
       * 5. Render the `editableData` returned from `useForm`
       */}
      <h1>{editableData.title}</h1>
      <p>{editableData.body}</p>
      <EditButton />
    </section>
  )
}

//...
```

## Edit the data

Head back to http://localhost:3000, enable the CMS, open the sidebar and try to update the content. You should now be able to edit the title and body copy on the demo!

![tina-tutorial-edit-data](/img/getting-started/edit-data.png)

## Form Configuration

To gain a little clarity into how this is working, let's look closer at the `formConfig`.

`useForm` needs a [form configuration object](/docs/plugins/forms#form-configuration) with properties that determine how the form behaves on load and save, what fields are available, and other metadata.

While there are other form config properties, the key ones to examine in the example above are `id`, `fields`, `initialValues`, and `onSubmit`. We'll look closely at each of these now.

### _id_

The `id` is unique identifier for the form. For this simple example the `id` is hard-coded, however it is good practice to set `id` to as a unique identifier from the content being edited.

### _fields_

> The `fields` array is comprised of [field definitions](/docs/plugins/fields#field-definition). All fields share a common [base configuration](docs/plugins/fields#field-config). Field definitions need at least two properties: a `name` or the path to the data and a `component` to edit that data from.

Reviewer 2: ```suggestion
The `fields` array is comprised of [field definitions](/docs/plugins/fields#field-definition). All fields share a common [base configuration](docs/plugins/fields#field-config). Field definitions need at least two properties: 

 - a `name`, indicating the path to access the data from the values object, and
 - a `component` that displays the interface for editing the data.
```

You will be working with fields a lot in Tina. To get more familiar, try to adjust the `label` property or add a new field to the array. Reference [the documentation](/docs/plugins/fields) for a full list of default field plugins.

### _initialValues_

> `initialValues` contains the data used to populate the form. If you do need to load data asynchronously on form creation, you'd use a function called `loadInitialValues` (we'll get to that soon).

Reviewer 2: ```suggestion
`initialValues` contains the data used to populate the form. If you need to load data asynchronously on form creation, you can instead use a function called `loadInitialValues` (we'll get to that soon).
```
